go/arrow/compare_test.go

```diff
@@ -0,0 +1,245 @@
+					Field{Name: "f2", Type: PrimitiveTypes.Uint32, Nullable: true},
+				},
+				index: map[string]int{"f1": 0, "f2": 0},
+			},
+			false, true,
+		},
+		{
+			&StructType{
+				fields: []Field{
+					Field{Name: "f1", Type: PrimitiveTypes.Uint16, Nullable: true},
+					Field{Name: "f2", Type: PrimitiveTypes.Float32, Nullable: false},
+				},
+				index: map[string]int{"f1": 0, "f2": 1},
+			},
+			&StructType{
+				fields: []Field{
+					Field{Name: "f1", Type: PrimitiveTypes.Uint16, Nullable: true},
+					Field{Name: "f2", Type: PrimitiveTypes.Float32, Nullable: false},
+				},
+				index: map[string]int{"f1": 0, "f2": 1},
+			},
+			true, false,
+		},
+		{
+			&StructType{
+				fields: []Field{
+					Field{Name: "f1", Type: PrimitiveTypes.Uint16, Nullable: true},
+					Field{Name: "f2", Type: PrimitiveTypes.Float32, Nullable: false},
+				},
+				index: map[string]int{"f1": 0, "f2": 1},
+			},
+			&StructType{
+				fields: []Field{
+					Field{Name: "f1", Type: PrimitiveTypes.Uint16, Nullable: true},
+					Field{Name: "f2", Type: PrimitiveTypes.Float32, Nullable: false},
+				},
+				index: map[string]int{"f1": 0, "f2": 1},
+			},
+			true, false,
+		},
+		{
+			&StructType{
+				fields: []Field{
+					Field{Name: "f1", Type: PrimitiveTypes.Uint16, Nullable: true},
+					Field{Name: "f2", Type: PrimitiveTypes.Float32, Nullable: false},
+				},
+				index: map[string]int{"f1": 0, "f2": 1},
+				meta:  MetadataFrom(map[string]string{"k1": "v1"}),
+			},
+			&StructType{
+				fields: []Field{
+					Field{Name: "f1", Type: PrimitiveTypes.Uint16, Nullable: true},
+					Field{Name: "f2", Type: PrimitiveTypes.Float32, Nullable: false},
+				},
+				index: map[string]int{"f1": 0, "f2": 1},
+				meta:  MetadataFrom(map[string]string{"k1": "v1"}),
+			},
+			true, true,
+		},
+		{
+			&StructType{
+				fields: []Field{
+					Field{Name: "f1", Type: PrimitiveTypes.Uint32, Nullable: true},
+				},
+				index: map[string]int{"f1": 0},
+				meta:  MetadataFrom(map[string]string{"k1": "v1"}),
+			},
+			&StructType{
+				fields: []Field{
+					Field{Name: "f1", Type: PrimitiveTypes.Uint32, Nullable: true},
+				},
+				index: map[string]int{"f1": 0},
+				meta:  MetadataFrom(map[string]string{"k1": "v2"}),
+			},
+			true, false,
+		},
+		{
+			&StructType{
+				fields: []Field{
+					Field{Name: "f1", Type: PrimitiveTypes.Uint16, Nullable: true, Metadata: MetadataFrom(map[string]string{"k1": "v1"})},
+					Field{Name: "f2", Type: PrimitiveTypes.Float32, Nullable: false},
+				},
+				index: map[string]int{"f1": 0, "f2": 1},
+			},
+			&StructType{
+				fields: []Field{
+					Field{Name: "f1", Type: PrimitiveTypes.Uint16, Nullable: true, Metadata: MetadataFrom(map[string]string{"k1": "v2"})},
+					Field{Name: "f2", Type: PrimitiveTypes.Float32, Nullable: false},
+				},
+				index: map[string]int{"f1": 0, "f2": 1},
+			},
+			false, true,
+		},
+	}
+
+	for _, test := range tests {
+		got := TypeEquals(test.left, test.right, test.checkMetadata)
```

Comment:
perhaps use a sub-test?
_ie:_
```go
for _, test := range tests {
    t.Run("", func(t*testing.T) {
         ...
    })
}
```
and replace the use of `t.Errorf(...)` with `t.Fatalf(...)`.